instrumentation/elasticsearch/elasticsearch-rest-7.0/library/src/main/java/io/opentelemetry/instrumentation/elasticsearch/rest/v7_0/RestClientWrapper.java

```diff
@@ -0,0 +1,186 @@
+                      && Response.class == method.getReturnType()) {
+                    Request request = (Request) args[0];
+                    Context parentContext = Context.current();
+                    ElasticsearchRestRequest otelRequest =
+                        ElasticsearchRestRequest.create(request.getMethod(), request.getEndpoint());
+                    if (!instrumenter.shouldStart(parentContext, otelRequest)) {
+                      return method.invoke(target, args);
+                    }
+
+                    Response response = null;
+                    Throwable throwable = null;
+                    Context context = instrumenter.start(parentContext, otelRequest);
+                    try (Scope scope = context.makeCurrent()) {
+                      response = (Response) method.invoke(target, args);
+                    } catch (Throwable exception) {
+                      throwable = exception;
+                    } finally {
+                      instrumenter.end(context, otelRequest, response, throwable);
+                    }
+
+                    return response;
+                  } else if ("performRequestAsync".equals(method.getName())
+                      && args.length == 2
+                      && args[0] instanceof Request
+                      && args[1] instanceof ResponseListener) {
+
+                    Request request = (Request) args[0];
+                    ResponseListener responseListener = (ResponseListener) args[1];
+                    Context parentContext = Context.current();
+                    ElasticsearchRestRequest otelRequest =
+                        ElasticsearchRestRequest.create(request.getMethod(), request.getEndpoint());
+                    if (!instrumenter.shouldStart(parentContext, otelRequest)) {
+                      return method.invoke(target, args);
+                    }
+
+                    Throwable throwable = null;
+                    Context context = instrumenter.start(parentContext, otelRequest);
+                    args[1] =
+                        new RestResponseListener(
+                            responseListener, parentContext, instrumenter, context, otelRequest);
+                    try (Scope scope = context.makeCurrent()) {
+                      return method.invoke(target, args);
+                    } catch (Throwable exception) {
+                      throwable = exception;
+                    } finally {
+                      if (throwable != null) {
+                        instrumenter.end(context, otelRequest, null, throwable);
+                      }
+                      // span ended in RestResponseListener
+                    }
+                  }
+
+                  // delegate to wrapped RestClient
+                  return method.invoke(target, args);
+                }))
+        .make()
+        .load(RestClient.class.getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
```

Comment:
Is there any reason for loading the proxy into the `RestClient`-classloader?
If the proxy was coded manually, it would be in the classloader of the `RestClientWrapper`.

I'm thinking that loading into the `RestClient`-classloader could cause issues if the `RestClient` is loaded by a parent classloader due to the added `instrumenter` field. The type `Instrumenter` might not be visible from there, resulting in the proxy class not loading correctly.

Reply:
`RestClient` has a package private constructor, because of that we need to put the proxy into the same class loader. I need to think about the `Instrumenter` a bit, probably it is enough to change it to `Supplier<Instrumenter>`.

Reply:
I see, that makes sense. Not sure if changing the type to `Supplier<Instrumenter>` would be enough, because IINM the type parameter is retained for access via reflection. You could just type the field to `Object` and cast manually when reading it.